python/src/dataset.rs

```diff
@@ -93,6 +94,101 @@ fn convert_schema(arrow_schema: &ArrowSchema) -> PyResult<Schema> {
+        let dataset: Py<Dataset> = dataset.extract()?;
+        let ds = dataset.borrow(on.py()).ds.clone();
+        // Either a single string, which we put in a vector or an iterator
+        // of strings, which we collect into a vector
+        let on = PyAny::downcast::<PyString>(on)
+            .map(|val| vec![val.to_string()])
+            .or_else(|_| {
+                let iterator = on.iter().map_err(|_| {
+                    PyValueError::new_err(
+                        "The `on` argument to merge_insert must be a str or iterable of str",
+                    )
+                })?;
+                let mut keys = Vec::new();
+                for key in iterator {
+                    keys.push(PyAny::downcast::<PyString>(key?)?.to_string());
+                }
+                PyResult::Ok(keys)
+            })?;
+
+        let mut builder = LanceMergeInsertBuilder::try_new(ds, on)
+            .map_err(|err| PyValueError::new_err(err.to_string()))?;
+
+        // We don't have do_nothing methods in python so we start with a blank slate
+        builder
+            .when_matched(WhenMatched::DoNothing)
+            .when_not_matched(WhenNotMatched::DoNothing);
+
+        Ok(Self { builder, dataset })
+    }
+
+    pub fn when_matched_update_all(mut slf: PyRefMut<Self>) -> PyResult<PyRefMut<Self>> {
+        slf.builder.when_matched(WhenMatched::UpdateAll);
+        Ok(slf)
+    }
+
+    pub fn when_not_matched_insert_all(mut slf: PyRefMut<Self>) -> PyResult<PyRefMut<Self>> {
+        slf.builder.when_not_matched(WhenNotMatched::InsertAll);
+        Ok(slf)
+    }
+
+    pub fn when_not_matched_by_source_delete<'a>(
+        mut slf: PyRefMut<'a, Self>,
+        expr: Option<&str>,
+    ) -> PyResult<PyRefMut<'a, Self>> {
+        let new_val = if let Some(expr) = expr {
+            let dataset = slf.dataset.borrow(slf.py());
+            WhenNotMatchedBySource::delete_if(&dataset.ds, expr)
+                .map_err(|err| PyValueError::new_err(err.to_string()))?
+        } else {
+            WhenNotMatchedBySource::Delete
+        };
+        slf.builder.when_not_matched_by_source(new_val);
+        Ok(slf)
+    }
+
+    pub fn execute(&mut self, new_data: &PyAny) -> PyResult<()> {
+        let py = new_data.py();
+
+        let new_data: Box<dyn RecordBatchReader + Send> = if new_data.is_instance_of::<Scanner>() {
+            let scanner: Scanner = new_data.extract()?;
+            Box::new(
+                RT.block_on(Some(py), scanner.to_reader())?
+                    .map_err(|err| PyValueError::new_err(err.to_string()))?,
+            )
+        } else {
+            Box::new(ArrowArrayStreamReader::from_pyarrow(new_data)?)
+        };
+
+        let job = self
+            .builder
+            .try_build()
+            .map_err(|err| PyValueError::new_err(err.to_string()))?;
+
+        let new_self = RT
+            .block_on(None, job.execute_reader(new_data))?
```

Comment:
To avoid runtime-within-a-runtime issues, we should use `spawn` here instead:

```suggestion
        let new_self = RT
            .spawn(None, async move { job.execute_reader(new_data).await })?
```

Reply:
Regrettably, this leads to deadlock.

First the python thread spawns a non-blocking task to run `execute_reader`.
Then the python thread blocks on `rx`.
During `execute_reader` we call into `LanceReader::next` (e.g. to get the batch from `new_data`)
This leads to a call to `spawn`
That call to spawn launches a non-blocking task to read the next batch.
That call to spawn then blocks on `rx` (but we are no longer on a python thread!)

Reply:
I managed to make this work with `spawn` but I had to change `reader_to_stream`.  This method is used by `write_dataset`, `append`, etc. to convert from a synchronous reader to an asynchronous stream.  Previously, it did this with `stream::iter` which just calls `next()` and blocks the async thread while it waits for the next item.  However, I don't think this is a safe thing to do since `next()` in this case could be quite expensive.

In particular, the deadlock I was getting was because `next()` was calling a python function, which in turn called back into rust (to read from the source dataset) which then immediately deadlocked because it couldn't acquire the async thread.

The change I made was to change `reader_to_stream` to use a special iterator->stream adapter which makes each call to `next()` using `spawn_blocking`.

Reply:
Oh I think you solved what I was hitting in https://github.com/lancedb/lance/pull/1869 then. Cool!

Reply:
Hmm, seems I introduced a deadlock in test_fragment.py now :confused:

Reply:
Ok, the problem was that `test_fragment` was using multiprocessing with fork (which tokio doesn't support) so not actually a bug this time.